File: lib/utilities/lib/debug.js
/* 
 * Factory+ NodeJS Utilities
 * Debugging / logging support.
 * Copyright 2022 AMRC.
 */

import util from "util";

export class Debug {
    constructor (opts) {
        opts ??= {};
        const verb = opts.verbose 
            ?? process.env.VERBOSE
            ?? opts.default
            ?? "";

        this.levels = new Set();
        this.suppress = new Set();
        this.verbose = false;

        for (const lev of verb.split(",")) {
            if (lev == "1" || lev == "ALL")
                this.verbose = true;
            else if (lev.startsWith("!"))
                this.suppress.add(lev.slice(1))
            else
                this.levels.add(lev);
        }
    }

    log (level, msg, ...args) {
        const want = this.verbose || this.levels.has(level);
        if (!want || this.suppress.has(level))
            return;
        
        const out = util.format(msg, ...args);
        const spc = " ".repeat(Math.max(0, 8 - level.length));
        console.log(`${level}${spc} : ${out}`);
    }

    bound (level) {
        return this.log.bind(this, level);
    }
}

File: lib/py-edge-driver/src/amrc/factoryplus/edge_driver/debug.py
# Copyright (c) University of Sheffield AMRC 2025.

import logging
from typing import Dict, Any, Optional, Callable, Type, Mapping, Set

class Debug:
    def __init__(self, levels: str):
        """
        Initialize Debug class with optional verbosity settings.

        Args:
            levels: Comma-separated string of log levels to enable
        """
        self.levels = set()
        self.suppress = set()
        self.verbose = False

        levels = levels or ""

        for level in levels.split(","):
            if level == "1" or level == "ALL":
                self.verbose = True
            elif level.startswith("!"):
                self.suppress.add(level[1:])
            elif level:
                self.levels.add(level)

    def log(self, level: str, msg: str, *args) -> None:
        """
        Log a message if the level is enabled.

        Args:
            level: Log level name
            msg: Message format string
            *args: Arguments for message formatting
        """
        want = self.verbose or level in self.levels
        if not want or level in self.suppress:
            return

        try:
            out = msg % args if args else msg
        except TypeError:
            out = f"{msg} {' '.join(str(arg) for arg in args)}"

        spc = " " * max(0, 8 - len(level))
        print(f"{level}{spc} : {out}")

    def bound(self, level: str) -> Callable:
        """
        Return a bound logging function for the given level.

        Args:
            level: Log level to bind to

        Returns:
            Callable function that logs at the specified level
        """
        def bound_log(msg, *args):
            self.log(level, msg, *args)
        return bound_log
